File: Python/bigtable/write_bigtable.py
#  Copyright 2023 Google LLC
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

# standard libraries
import logging

# third party libraries
import apache_beam as beam
from apache_beam import Create, Map
from apache_beam.io.gcp.bigtableio import WriteToBigTable
from apache_beam.options.pipeline_options import PipelineOptions


class BigtableOptions(PipelineOptions):
    @classmethod
    def _add_argparse_args(cls, parser):
        # Add a command line flag to be parsed along
        # with other normal PipelineOptions
        parser.add_argument(
            "--project_id",
            required=True,
            help="Project ID"
        )
        parser.add_argument(
            "--instance_id",
            required=True,
            help="Cloud Bigtable instance ID"
        )
        parser.add_argument(
            "--table_id",
            required=True,
            help="Cloud Bigtable table ID"
        )


def run():
    """
    This pipeline shows how to write to Cloud Bigtable.
    """

    options = BigtableOptions()

    with beam.Pipeline(options=options) as p:
        # Create sample data
        elements = [
            (1, "Charles, 1995, USA"),
            (2, "Alice, 1997, Spain"),
            (3, "Bob, 1995, USA"),
            (4, "Amanda, 1991, France"),
            (5, "Alex, 1999, Mexico"),
            (6, "Eliza, 2000, Japan")
        ]

        output = (
            p
            | "Create elements" >> Create(elements)
            | "Map to BigTable Row" >> Map(make_bigtable_row)
            | "Write to BigTable" >> WriteToBigTable(
                project_id=options.project_id,
                instance_id=options.instance_id,
                table_id=options.table_id
            )
        )


def make_bigtable_row(element):
    """
    Converts a given input element into a BigTable DirectRow object.
    """

    # third party libraries
    from google.cloud.bigtable.row import DirectRow
    from datetime import datetime

    # Example transformation to BigTable DirectRow

    index = element[0]
    row_fields = element[1].split(", ")
    column_family_id = "cf1"
    key = "beam_key%s" % ("{0:07}".format(index))
    direct_row = DirectRow(row_key=key)
    for column_id, value in enumerate(row_fields):
        direct_row.set_cell(
            column_family_id,
            ("field%s" % column_id).encode("utf-8"),
            value,
            datetime.now()
        )
    return direct_row


if __name__ == "__main__":
    logging.getLogger().setLevel(logging.INFO)
    run()

File: Java/src/main/java/bigtable/WriteBigTable.java
/*
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package bigtable;

import com.google.bigtable.v2.Mutation;
import com.google.common.collect.ImmutableList;
import com.google.common.primitives.Ints;
import com.google.protobuf.ByteString;
import org.apache.beam.runners.dataflow.options.DataflowPipelineOptions;
import org.apache.beam.sdk.Pipeline;
import org.apache.beam.sdk.io.gcp.bigtable.BigtableIO;
import org.apache.beam.sdk.options.Default;
import org.apache.beam.sdk.options.Description;
import org.apache.beam.sdk.options.PipelineOptionsFactory;
import org.apache.beam.sdk.transforms.Create;
import org.apache.beam.sdk.transforms.DoFn;
import org.apache.beam.sdk.transforms.ParDo;
import org.apache.beam.sdk.transforms.windowing.GlobalWindow;
import org.apache.beam.sdk.values.KV;
import org.checkerframework.checker.nullness.qual.Nullable;
import org.joda.time.Instant;

import java.math.BigInteger;
import java.util.stream.IntStream;

public class WriteBigTable {
    /*
    TODO

    This pipeline needs a BigTable instance and tables, in this example we use
    quickstart examples:
    https://cloud.google.com/bigtable/docs/quickstart-cbt

    Two families need to be created, "prime" and "not prime"
    $ cbt createfamily my-table prime
    $ cbt createfamily my-table not-prime
     */

    /**
     * Pipeline options to be passed via the command line.
     */
    public interface WriteBigTableOptions extends DataflowPipelineOptions {
        @Description("Instance ID")
        @Default.String("quickstart-instance")
        String getInstance();

        void setInstance(String value);

        @Description("Table ID")
        @Default.String("my-table")
        String getTable();

        void setTable(String value);

        @Nullable
        @Description("Project ID")
        String getProjectBT();

        void setProjectBT(String value);
    }

    public static void main(String[] args) {

        // Parse pipeline options from the command line.
        WriteBigTableOptions options = PipelineOptionsFactory.fromArgs(args).withValidation().as(WriteBigTableOptions.class);

        // Use Dataflow project if ProjectBT pipeline option is not passed.
        String project = (options.getProjectBT() == null) ? options.getProject() : options.getProjectBT();

        // Create the pipeline.
        Pipeline p = Pipeline.create(options);

        int[] rangeIntegers = IntStream.range(0, 100).toArray();
        Iterable<Integer> elements = Ints.asList(rangeIntegers);

        p
                // Add a source that uses the hard-coded values (integers from 0 to 99).
                .apply(Create.of(elements))
                // Add the CreateMutations transform that groups the values into bundles
                // of Bigtable "mutations".
                .apply("Group in Mutations", ParDo.of(new CreateMutations()))
                // The Bigtable sink takes as input a PCollection<KV<ByteString, Iterable<Mutation>>>,
                // where the ByteString is the key of the row being mutated, and each Mutation
                // represents an idempotent transformation to that row.
                .apply(BigtableIO.write()
                        .withInstanceId(options.getInstance())
                        .withProjectId(project)
                        .withTableId(options.getTable()));

        // Execute the pipeline.
        p.run();
    }

    public static class CreateMutations extends DoFn<Integer, KV<ByteString, Iterable<Mutation>>> {
        public ImmutableList.Builder<Mutation> mutations;

        @ProcessElement
        public void processElement(ProcessContext c) {
            BigInteger b = new BigInteger(c.element().toString());
            String isPrime = b.isProbablePrime(1) ? "prime" : "not-prime";

            Mutation.SetCell setCell =
                Mutation.SetCell.newBuilder()
                    .setFamilyName(isPrime)
                    .setColumnQualifier(toByteString(c.element().toString()))
                    .setValue(toByteString("value-" + c.element()))
                    .setTimestampMicros(Instant.now().getMillis() * 1000)
                    .build();
            this.mutations.add(Mutation.newBuilder().setSetCell(setCell).build());
        }

        @StartBundle
        public void startBundle() {
            this.mutations = ImmutableList.builder();
        }

        @FinishBundle
        public void finishBundle(FinishBundleContext c) {
            KV<ByteString, Iterable<Mutation>> row = KV.of(toByteString("numbers"), this.mutations.build());
            c.output(row, Instant.now(), GlobalWindow.INSTANCE);
        }

        private static ByteString toByteString(String value) {
            return ByteString.copyFromUtf8(value);
        }
    }
}
